### PR TITLE
New options for encoding Go strings to and from CBOR byte strings

### PR DESCRIPTION
**Pull request by benluddy**
#### PR Was Proposed and Welcomed in Currently Open Issue

- [x] This PR was proposed and welcomed by maintainer(s) in issue #446
- [x] Closes or Updates Issue #446

#### Checklist (for code PR only, ignore for docs PR)

- [x] Include unit tests that cover the new code
- [x] Pass all unit tests 
- [x] Pass all 18 ci linters (golint, gosec, staticcheck, etc.)
- [x] Sign each commit with your real name and email.  
      Last line of each commit message should be in this format:  
      Signed-off-by: Firstname Lastname <firstname.lastname@example.com>
- [x] Certify the Developer's Certificate of Origin 1.1
      (see next section).

#### Certify the Developer's Certificate of Origin 1.1

- [x] By marking this item as completed, I certify 
      the Developer Certificate of Origin 1.1.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
660 York Street, Suite 102,
San Francisco, CA 94110 USA

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

File: encode.go [hunk 1]
```diff
@@ -156,6 +156,27 @@ func (sm SortMode) valid() bool {
...
+// StringType is a CBOR string type.
+type StringType int
+
+const (
+	// TextStringType is the CBOR text string type (major type 3).
+	TextStringType StringType = iota
+
+	// ByteStringType is the CBOR byte string type (major type 2).
+	ByteStringType
+)
```
Comment (fxamacker) [inline encode.go:168]: Maybe use similar naming convention as existing options like `TimeMode`, etc.

```suggestion
// StringMode specifies how to encode Go string values.
type StringMode int

const (
	// StringToTextString encodes Go string to CBOR text string (major type 3).
	StringToTextString StringMode = iota

	// StringToByteString encodes Go string to CBOR byte string (major type 2).
	StringToByteString
)
```

File: encode.go [hunk 2]
```diff
@@ -326,6 +347,9 @@ type EncOptions struct {
 
 	// TagsMd specifies whether to allow CBOR tags (major type 6).
 	TagsMd TagsMode
+
+	// StringType specifies which CBOR string major type to use when encoding Go strings.
+	StringType StringType
```
Comment (fxamacker) [inline encode.go:352]:
```suggestion
        // String specifies which CBOR type to use when encoding Go strings.
	// - CBOR text string (major type 3) is default
        // - CBOR byte string (major type 2)
	String StringMode
```

File: decode.go [hunk 3]
```diff
@@ -412,6 +412,23 @@ func (bidm BigIntDecMode) valid() bool {
...
 
+// ByteStringToStringMode specifies the behavior when decoding a CBOR byte string into a Go string.
+type ByteStringToStringMode int
+
+const (
+	// ByteStringToStringError generates an error on an attempt to decode a CBOR byte string into a Go string.
+	ByteStringToStringError ByteStringToStringMode = iota
+
+	// ByteStringToStringAllow permits decoding a CBOR byte string into a Go string.
+	ByteStringToStringAllow
```
Comment (fxamacker) [inline decode.go:423]: Maybe rename to use similar naming convention as existing options like `MapKeyByteStringMode`.

```suggestion
	// ByteStringToStringForbidden forbids decoding CBOR byte string into Go string by
	// returning UnmarshalTypeError.
	ByteStringToStringForbidden ByteStringToStringMode = iota

	// ByteStringToStringAllowed allows decoding a CBOR byte string into a Go string.
	ByteStringToStringAllowed
```

File: decode.go [hunk 4]
```diff
@@ -1279,7 +1317,29 @@ func (d *decoder) parse(skipSelfDescribedTag bool) (interface{}, error) { //noli
 		return nValue, nil
 
 	case cborTypeByteString:
-		return d.parseByteString(), nil
+		switch d.dm.defaultByteStringType {
+		case nil, reflect.TypeOf([]byte(nil)):
```
Comment (fxamacker) [inline decode.go:1321]: nit: maybe create a global var for `reflect.TypeOf([]byte(nil))` so it can be reused.

File: decode.go [hunk 5]
```diff
@@ -1279,7 +1317,29 @@ func (d *decoder) parse(skipSelfDescribedTag bool) (interface{}, error) { //noli
...
+			b, copied := d.parseByteString()
+			if copied {
+				return b, nil
+			}
+			clone := make([]byte, len(b))
+			copy(clone, b)
+			return clone, nil
+		case reflect.TypeOf(""):
```
Comment (fxamacker) [inline decode.go:1329]: nit: maybe create a global var for `reflect.TypeOf("")` so it can be reused.

File: decode.go [hunk 6]
```diff
@@ -1376,15 +1436,16 @@ func (d *decoder) parse(skipSelfDescribedTag bool) (interface{}, error) { //noli
...
 }
 
-// parseByteString parses CBOR encoded byte string.  It returns a byte slice
-// pointing to a copy of parsed data.
-func (d *decoder) parseByteString() []byte {
+// parseByteString parses a CBOR encoded byte string. The returned byte slice
+// may be backed directly by the input. The second return value will be true if
+// and only if the slice is backed by a copy of the input. Callers are
+// responsible for making a copy if necessary.
+func (d *decoder) parseByteString() ([]byte, bool) {
```
Comment (fxamacker) [inline decode.go:1443]: Thanks for optimizing this! :+1:

File: decode.go [hunk 7]
```diff
@@ -2184,18 +2245,31 @@ func fillFloat(t cborType, val float64, v reflect.Value) error {
...
 
-func fillByteString(t cborType, val []byte, v reflect.Value) error {
+func fillByteString(t cborType, val []byte, shared bool, v reflect.Value, bsts ByteStringToStringMode) error {
 	if reflect.PtrTo(v.Type()).Implements(typeBinaryUnmarshaler) {
 		if v.CanAddr() {
 			v = v.Addr()
 			if u, ok := v.Interface().(encoding.BinaryUnmarshaler); ok {
+				// The contract of BinaryUnmarshaler forbids
+				// retaining the input bytes, so no copying is
+				// required even if val is shared.
```
Comment (fxamacker) [inline decode.go:2255]: Thanks for adding this comment! :+1:

**Review by fxamacker — COMMENTED** [review]
This is great! :+1:  Thanks for adding this useful feature and also for optimizing byte string decoding while at it!  

I left some minor suggestions about naming conventions used by existing options, etc.

**Comment by benluddy** [comment]
Thanks @fxamacker, I've implemented all of your suggestions. Please take another look!

**Review by fxamacker — APPROVED** [review]
LGTM!  :+1:  Thanks again for this PR!

**Pull request merged**